### PR TITLE
Fix encoding for OSX keychain

### LINKED ISSUE #1283: MacOS Keychain request prompts for "M" (or similar)
image

We should investigate why this is being listed as a single letter which isn't really helpful for understanding what STS is trying to access. Ideally it would have a more friendly name like "Connection Passwords" or similar.

### PR DESCRIPTION
**Pull request by cheenamalhotra**
Fixes microsoft/azuredatastudio#22385 

The prompt now provides correct key information (fixed in last commit)

_P.S. With the new change to encoding, it did seem credentials will be lost. I've added code to migrate credentials to new encoding so users can stay unimpacted._

NOTE: This PR only fixes the encoding of credentials and prompt, and does not change the frequency of prompts.

### CODE REVIEW

**Review by Charles-Gagnon — COMMENTED** [review]
Does this affect the existing stored passwords at all? Or is it essentially just a UI fix?

**Comment by cheenamalhotra** [comment]
> Does this affect the existing stored passwords at all? Or is it essentially just a UI fix?

The interop APIs continue to transfer and read in Unicode... https://github.com/microsoft/sqltoolsservice/blob/0d4c9c9a622fd5422c3ab55c7e9fb67252c4c03e/src/Microsoft.SqlTools.Credentials/Credentials/OSX/Interop.Security.cs#L15

I haven't used MacOS for dev/testing until now so I don't have much keys to verify this.. I only tested it with Azure MFA by disabling 'noSystemKeyChain' with ADAL to get this prompt (and it hasn't caused me any issues that I could make note of). How else can I ensure this API gets called from ADS (if I want to test before/after)? Does SQL Login save password cover keychain access?

Alternatively, if you can help validate this branch if you have saved passwords in keychain, that would be great :)

**Comment by Charles-Gagnon** [comment]
I don't have a Mac unfortunately - @alanrenmsft could you help?

This API should be called every time a SQL login password is stored (or I guess with Azure auth tokens as well given your example above)

An easy way to test this though would be to:

1. Open up ADS stable and save a password
2. Verify that it saved and is loaded correctly
3. Build local dev with your changes here and then launch dev ADS with those changes and verify that you can still access the saved password

**Review by Charles-Gagnon — APPROVED** [review]
Change looks good assuming there's no issues with loading existing saved passwords.

**Review by kburtram — APPROVED** [review]
LGTM, I'll verify on my mac later this evening.

**Comment by alanrenmsft** [comment]
let me test it on OSX-ARM64

**Comment by alanrenmsft** [comment]
modifying the ADS app on osx-arm64 doesn't work since it has more strict security check, I had to produce a new build in order to test it.

**Comment by alanrenmsft** [comment]
@cheenamalhotra how did you test the changes? I produced a new build with your changes and it is still showing me 'M' in the prompt.

This is the build I used, mssql is pointing to 5.0.0.1 version of sts, which is created manually by me with the files produced by the PR validation build in the PR. unfortunately, the macOS job failed, I've restarted it, could you please try again when it is done?

**Comment by cheenamalhotra** [comment]
I tested with the `ADS_SQLTOOLSERVICE` env variable (added in `launch.json`) pointing to the output directory of STS published artifacts for `osx-arm64` target. Seemed to work fine without any strict checks.

**Comment by cheenamalhotra** [comment]
Synced offline with @alanrenmsft and seems we still are sending some Unicode strings as Credential ID when storing password, will have to look into it as well.

**Comment by cheenamalhotra** [comment]
@alanrenmsft Please take a look again when you get a chance, I figured out the charset encoding was also contributing to the issue and have addressed it.

**Comment by cheenamalhotra** [comment]
> @cheenamalhotra just to clarify: the previously saved passwords will not be accessible, right? also, please update the screenshots

They will be migrated to new encoding when accessed, so you wouldn't notice missing passwords when running ADS.

File: src/Microsoft.SqlTools.Credentials/Credentials/OSX/OSXCredentialStore.cs [hunk 1]
```diff
@@ -96,10 +96,40 @@ private KeyChainItemHandle LookupKeyChainItem(string credentialId)
...
+                {
+                    var handle = new KeyChainItemHandle(item, passwordPtr, passwordLength);
+                    // Migrate credential to 'Auto' encoding.
+                    if (handle != null)
+                    {
+                        var saveResult = this.Save(credential: new Credential(credentialId, handle.Password));
+                        if (saveResult)
+                        {
+                            // Safe to delete old password now.
+                            this.DeletePassword(credentialId);
```
Comment (Charles-Gagnon) [inline src/Microsoft.SqlTools.Credentials/Credentials/OSX/OSXCredentialStore.cs:127]: Isn't this going to delete the new password? This should be calling

```csharp
Interop.Security.OSStatus status = Interop.Security.SecKeychainItemDelete(handle);
```

instead (right now it's doing a new lookup of the credential). You could also make new `LookupKeyChainItemOld` and `DeletePasswordOld` helper methods if you wanted (or probably in that case would be to just add a parameter to the existing ones to specify whether it's using the old implementations)

Comment (cheenamalhotra) [reply src/Microsoft.SqlTools.Credentials/Credentials/OSX/OSXCredentialStore.cs:127]: Good catch! Will address it.

File: src/Microsoft.SqlTools.Credentials/Credentials/OSX/OSXCredentialStore.cs [hunk 2]
```diff
@@ -96,10 +96,40 @@ private KeyChainItemHandle LookupKeyChainItem(string credentialId)
...
+                    credentialId,
+                    0,
+                    null,
+                    out passwordLength,
+                    out passwordPtr,
+                    out item);
+#pragma warning restore 0612
+                if (status == Interop.Security.OSStatus.ErrSecSuccess)
+                {
+                    var handle = new KeyChainItemHandle(item, passwordPtr, passwordLength);
```
[the remark this answers is not in the document]
Comment (cheenamalhotra) [reply src/Microsoft.SqlTools.Credentials/Credentials/OSX/OSXCredentialStore.cs:119]: This `handle` is returned later.

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/Microsoft.SqlTools.Credentials/Credentials/OSX/OSXCredentialStore.cs:119]: Oh whoops, got mixed up with the collapsed code sections...

**Comment by cheenamalhotra** [comment]
I need to give it some more time it seems, will update when I get it working with proper new entries in keychain.

**Comment by cheenamalhotra** [comment]
Verified ✅ Secrets are now migrated to UTF8 encoding.

Users will get prompted once when reading old credential with the old unicode credential ID - but soon after migration, correct encoding will be used to access credentials.

**Pull request merged**